### DISCUSSION TITLE
Deploy stucks on message about RubyZip 3.0 coming

### DISCUSSION THREAD

**Discussion by LawfulEvilRaccoon** [basecamp/kamal#747] [Q&A]
Hello!

I'm new to Rails and I'm tying to deploy Rails project to my VPS via Kamal for the first time. And the process stucks on warning about that RubyZip 3.0 is coming. I got no error messages, just this:

```
 DEBUG [0729c69a] 	#29 668.9 Bundled gems are installed into `/usr/local/bundle`
 DEBUG [0729c69a] 	#29 668.9 Post-install message from rubyzip:
 DEBUG [0729c69a] 	#29 668.9 RubyZip 3.0 is coming!
 DEBUG [0729c69a] 	#29 668.9 **********************
 DEBUG [0729c69a] 	#29 668.9 
 DEBUG [0729c69a] 	#29 668.9 The public API of some Rubyzip classes has been modernized to use named
 DEBUG [0729c69a] 	#29 668.9 parameters for optional arguments. Please check your usage of the
 DEBUG [0729c69a] 	#29 668.9 following classes:
 DEBUG [0729c69a] 	#29 668.9   * `Zip::File`
 DEBUG [0729c69a] 	#29 668.9   * `Zip::Entry`
 DEBUG [0729c69a] 	#29 668.9   * `Zip::InputStream`
 DEBUG [0729c69a] 	#29 668.9   * `Zip::OutputStream`
 DEBUG [0729c69a] 	#29 668.9 
 DEBUG [0729c69a] 	#29 668.9 Please ensure that your Gemfiles and .gemspecs are suitably restrictive
 DEBUG [0729c69a] 	#29 668.9 to avoid an unexpected breakage when 3.0 is released (e.g. ~> 2.3.0).
 DEBUG [0729c69a] 	#29 668.9 See https://github.com/rubyzip/rubyzip for details. The Changelog also
 DEBUG [0729c69a] 	#29 668.9 lists other enhancements and bugfixes that have been implemented since
 DEBUG [0729c69a] 	#29 668.9 version 2.3.0.
```

No matter if the command in terminal was "kamal setup" or "kamal deploy", the process stucks right here every time. I've been waiting first time for about an hour, second time for two, but nothing more happened

What can be a reason for that behavior and how can I avoid this situation? What should I pay attention and/or fix to get my project deployed?

Thank you in advance for your help.

**Comment by clwy-cn** [comment]
I encountered this issue as well. I attempted deployment five times and each time, it got stuck at RubyZip 3.0. However, when I changed the Ruby version in the Dockerfile from 3.3.0 to 

```yml
ARG RUBY_VERSION=3.2.3
```

and also updated `.ruby-version` and `Gemfile` to `3.2.3`, then ran `kamal deploy` again, all the issues vanished. The deployment was very smooth and successful.

**Comment by LawfulEvilRaccoon** [comment]
Thank you for your reply, Dong. Maybe I'll try to "play" with ruby version
if I have the same issue when I have to deploy something next time.
However, this time I've just built a docker image of my project and
deployed it via private docker registry. So I don't think that problem was
related to Dockerfile and ruby version in it. More like the problem is
Kamal itself with ruby 3.3.0, as I had my issue on the same ruby version.

**Comment by matthee** [comment]
I encountered a similar issue, when using `docker buildx` to cross compile a rails application for i386 (I'm using docker directly). As far as I can tell, this is an issue related to bootsnap getting stuck in the `bootsnap precompile precompile --gemfile` command. 

Maybe related? https://github.com/Shopify/bootsnap/issues/415

**Reply by 23hp** [reply]
I was trying to build an arm image on a intel CPU MacBook, and encounter this issue, I solve this by changing to a ARM CPU Macbook, no code changes are needed

**Comment by ghaydarov** [comment]
Anyone has any solution to this? Encountering this now. Ruby version 3.3.2

**Comment by szluyufeng** [comment]
Same problem here. Any solution？

**Comment by kyrylo** [comment]
I'm experiencing the same issue. I'm on:

- macOS Ventura 13.2 (Intel CPU)
- Ruby 3.3.0
- Docker version 26.1.1, build 4cf5afa

**Comment by thermistor** [comment]
My kamal deploy was also hanging and reverting to ruby `3.2.4` from `3.3.3` solved it, thanks @clwy-cn.

**Comment by anchietajunior** [comment]
Kamal deploy/setup freezes for me everytime in this step:
Versions: 
   - Ruby 3.2.4 and Ruby 3.3.3
   - Docker 26.1.1
   -  MacOS Sonoma 14.5

```shell
 $ yarn build:css:compile && yarn build:css:prefix
 $ sass ./app/assets/stylesheets/application.bootstrap.scss:./app/assets/builds/application.css --no-source-map --load-path=node_modules
 ```
 
 ## Edit
 
 The problem for me was sass or bootstrap, i changed to --css tailwind and it worked.

**Comment by GrantBarry** [comment]
Same problem here when trying to build with Ruby 3.3.x (current latest version 3.3.4 and also has the problem).
Moving back to a Ruby 3.2.x resolves the problem (current latest version 3.2.5).

Once bundler has finished, the build doesn't carry on with the rest of the build process.

**Reply by jabebbo** [reply]
This also worked for me,
Reverted back from Ruby 3.3.6 to 3.2.6 using rbenv,
then installed latest Rails 8.0.0, Bundler version 2.5.23, Kamal version 2.3.0 
running on Apple M1 Max with docker desktop, 

# Configure the image builder.
builder:
  arch: amd64

**Comment by srbaker** [comment]
Same problem here.

**Comment by fedeth** [comment]
Same problem here

**Comment by GrantBarry** [comment]
Updated to Docker v4.34.0, and Ruby 3.3.5 and I still have the same problem.
Kamal is version 1.8.3
Bundler 2.5.18

If someone can point me in a direction, I can try help debug. I just don't know where to start.

Otherwise, what additional information would be helpful?

Thanks.

**Comment by GrantBarry** [comment]
When I press Control + C to interrupt the process, I get the following dump:


/Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/local.rb:57:in `join': Interrupt
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/local.rb:57:in `block in execute_command'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/3.3.0/open3.rb:540:in `popen_run'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/3.3.0/open3.rb:235:in `popen3'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/local.rb:44:in `execute_command'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/abstract.rb:148:in `block in create_command_and_execute'
	from <internal:kernel>:90:in `tap'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/abstract.rb:148:in `create_command_and_execute'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/abstract.rb:80:in `execute'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/build.rb:55:in `block (3 levels) in push'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/build.rb:55:in `chdir'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/build.rb:55:in `block (2 levels) in push'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/commander.rb:123:in `with_verbosity'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/build.rb:54:in `block in push'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/abstract.rb:31:in `instance_exec'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/backends/abstract.rb:31:in `run'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/sshkit-1.23.1/lib/sshkit/dsl.rb:10:in `run_locally'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/build.rb:36:in `push'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/build.rb:8:in `deliver'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor/command.rb:28:in `run'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor.rb:538:in `dispatch'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor/invocation.rb:116:in `invoke'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/main.rb:36:in `block in deploy'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/base.rb:97:in `print_runtime'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/lib/kamal/cli/main.rb:25:in `deploy'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor/command.rb:28:in `run'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor.rb:538:in `dispatch'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/thor-1.3.2/lib/thor/base.rb:584:in `start'
	from /Users/grant/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/kamal-1.8.3/bin/kamal:9:in `<top (required)>'
	from /Users/grant/.rbenv/versions/3.3.5/bin/kamal:25:in `load'
	from /Users/grant/.rbenv/versions/3.3.5/bin/kamal:25:in `<main>'

**Comment by olance** [comment]
In my case, changing Ruby versions did not help – however, the comment above about building on Intel for ARM got me to something working: in my case, I'm on an Intel MacPro and building for an Intel/AMD64 VPS.

So I added the following to my `deploy.yml` file to skip the ARM build:

```yaml
builder:
  local:
    arch: amd64
```

And it worked like a charm :)

**Comment by GrantBarry** [comment]
Well done Olivier, that was my problem!

I was missing that in my configuration and my builds work when I add this in. Plus, they build faster since it seems the default option builds for ARM and AMD ...

Thank you very much!!

**Reply by olance** [reply]
Glad it worked! :)

**Comment by dhh** [comment]
Kamal 2 will default to only building for the architecture specified in builder/arch rather than multiarch by default. This should mean very quick builds when you're building for the same architecture, but also faster builds even when you're doing it cross-architecture.

**Comment by stmpjmpr** [comment]
This still happens to me on Kamal 2.3.0, developing locally on Mac and deploying to arm64 on Hetzner, with the `builder` set to `arch: amd64`. I had to set a remote amd64 builder to get it to work.

**Comment by judy** [comment]
I had the same problem, and don't have it anymore... and wish I could tell you exactly what I did to fix it. 😬 I was using Docker more frequently years ago, and reinstalled it just to try out Kamal. New Rails 8 project, nothing weird in it. I opened up Docker Desktop, stopped all containers and deleted all images and everything. Stopped Docker Desktop and restarted it. Tried building again, and got past the end of the Bundler stage.

Last time I used Docker, I had to figure out how the buildx stuff works so I could build a Ruby gem for multiple architectures. So my guess is a bad build image.

**Comment by refacrot** [comment]
I still have the same issue. For me, it stuck after bundle install:
```
 ...
 DEBUG [cc381d37] 	#15 103.7 Installing rails 7.2.2
 DEBUG [cc381d37] 	#15 103.7 Installing turbo-rails 2.0.11
 DEBUG [cc381d37] 	#15 103.7 Installing tailwindcss-rails 3.0.0
 DEBUG [cc381d37] 	#15 103.9 Bundle complete! 15 Gemfile dependencies, 90 gems now installed.
 DEBUG [cc381d37] 	#15 103.9 Gems in the group 'development' were not installed.
 DEBUG [cc381d37] 	#15 103.9 Bundled gems are installed into `/usr/local/bundle`
 DEBUG [cc381d37] 	#15 103.9 Post-install message from i18n:
 DEBUG [cc381d37] 	#15 103.9 PSA: I18n will be dropping support for Ruby < 3.2 in the next major release (April 2025), due to Ruby's end of life for 3.1 and below (https://endoflife.date/ruby). Please upgrade to Ruby 3.2 or newer by April 2025 to continue using future versions of this gem.
 ```
I was waiting about 3 hours - no results.

Ruby 3.3.5
Rails 7.2.2
Kamal 2.3.0
M1 Max
arc: amd64

**UPDATE**: Fix found

Make sure you have VM setting in Docker set to this:
<img width="555" alt="image" src="https://github.com/user-attachments/assets/571d496c-d347-40d6-9e27-fb9ac9dabbb6">

I had `Docker VMM` enabled, and it didn't work. Now it works.

**Reply by hwrd** [reply]
Thank you!

Turning on the "Use Rosetta for x86_64/amd64 emulation on Apple Silicon" option solved the problem for me.

**Reply by morszczuk** [reply]
I can confirm that this solution worked for me as well. Thank you! 🙌

**Reply by erichstark** [reply]
Thanks, It still helps :D

**Reply by indigotechtutorials** [reply]
Yea but this breaks TailwindCSS deploys 😢

https://github.com/tailwindlabs/tailwindcss/issues/17728#issuecomment-2884340873

**Comment by pinzonjulian** [comment]
I was running into this using OrbStack on a Mac M3 Pro. The solution was to check the `Use Rosetta to run Intel code` checkbox. I don't even know why it was disabled in my machine.

<img width="545" alt="image" src="https://github.com/user-attachments/assets/c9bab558-9c5c-447e-8cf5-3ae81ee2aa9c" />

**Reply by dankimio** [reply]
for me, turning off this option helped, weird
OrbStack 2.0.1

**Comment by indigotechtutorials** [comment]
So the solution to fix it also breaks deploys with TailwindCSS.

I think it has something to do with the default in deploy being AMD64 and I'm on mac with Arm64 gonna experiment with this

<img width="1086" height="527" alt="Screenshot 2025-09-06 at 5 00 35 PM" src="https://github.com/user-attachments/assets/4a767fdc-6309-42b1-afb5-1a83d7af5ed6" />
